### PR TITLE
CUDA performance fixes

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
1. Added preallocated buffer to avoid cuda malloc/free
2. Added dedicated stream in cuda reduction kernel
3. Added reduce for multiple vectors

### CODE REVIEW

File: src/utils/mem_component.c [hunk 1]
```diff
@@ -9,13 +9,14 @@ static xccl_mem_component_t *mem_components[UCS_MEMORY_TYPE_LAST];
...
-    int    mt;
+    void          *handle;
+    char          *mem_comp_path;
+    size_t        mem_comp_path_len;
+    int           mt;
+    xccl_status_t status;
 
     mem_comp_path_len = strlen(components_path) + 32;
-    mem_comp_path = (char*)malloc(mem_comp_path_len);
+    mem_comp_path     = (char*)malloc(mem_comp_path_len);
```
Comment (vspetrov) [inline src/utils/mem_component.c:19]: check for NULL ,ret ERR_NO_MEMORY

File: src/utils/mem_component.h [hunk 2]
```diff
@@ -30,4 +43,16 @@ xccl_status_t xccl_mem_component_reduce(void *sbuf1, void *sbuf2, void *target,
                                         size_t count, xccl_dt_t dtype,
                                         xccl_op_t op, ucs_memory_type_t mem_type);
 
+/* 
+ * Perfroms reduction of multiple vectors and stores result to rbuf
+ * rbuf = sbuf1 + sbuf2{0} + sbuf2{1} + sbuf2{count-1}
+ * count  - number of vectros in sbuf2
```
Comment (vspetrov) [inline src/utils/mem_component.h]: vectros -> vectors
Perfroms -> Performs

File: src/utils/mem_component.c [hunk 3]
```diff
@@ -80,6 +107,31 @@ xccl_status_t xccl_mem_component_reduce(void *sbuf1, void *sbuf2, void *target,
...
 }
 
+xccl_status_t
+xccl_mem_component_reduce_multi(void *sbuf1, void *sbuf2, void *rbuf, size_t count,
+                                size_t size, size_t stride, xccl_dt_t dtype,
+                                xccl_op_t op, ucs_memory_type_t mem_type)
+{
+    int i;
+
+    assert(count > 1);
```
Comment (vspetrov) [inline src/utils/mem_component.c]: Why this assertion? count = 1 should be the case of radix=2 - you should have seen this assertion failure, no?
Imho, count = 1 should just work as is.

**Review by vspetrov — CHANGES_REQUESTED** [review]
Regarding the cache in mem_component:
lets merge it this way for now and then update to at least hcoll approach to be more generic.

File: src/utils/mem_component.c (hunk 3 above)
Comment (Sergei-Lebedev) [reply src/utils/mem_component.c]: count=1 will work, I just thought that there are no reasons to use multireduce in this case, but actually radix=2 is the case

Comment (Sergei-Lebedev) [reply src/utils/mem_component.c]: And we didn't see assert because it was release build. With --enable-debug and radix=2 it checks and fails

**Review by vspetrov — APPROVED** [review]

**Pull request merged**